examples/training/resnet50/main.cpp

```diff
@@ -0,0 +1,253 @@
+  // training.
+  Tensor Images(ElemKind::FloatTy, AllImagesDims);
+  Tensor Labels(ElemKind::FloatTy, AllLabelsDims);
+
+  llvm::outs() << "Zeroing Images/Labels memory.\n";
+  Images.zero();
+  Labels.zero();
+
+  /// The CIFAR file format is structured as one byte label in the range 0..9.
+  /// The label is followed by an image: 32 x 32 pixels, in RGB format. Each
+  /// color is 1 byte. The first 1024 red bytes are followed by 1024 of green
+  /// and blue. Each 1024 byte color slice is organized in row-major format.
+  /// The database contains 10000 images.
+  /// Size: (1 + (32 * 32 * 3)) * 10000 = 30730000.
+  llvm::outs() << "Assigning Images/Labels bytes.\n";
+  auto LabelsH = Labels.getHandle<>();
+  auto ImagesH = Images.getHandle<>();
+  float Scale = 1. / 255.0;
+  float Bias = 0;
+
+  /// CIFAR image dimensions are 32x32, ResNet50 expects 224x224.
+  /// Simple scaling would be one CIFAR pixel copied into 7x7 area
+  constexpr unsigned SCALE = 7;
+  for (unsigned N = 0; N < CIFAR_NUM_IMAGES; ++N) {
+    unsigned long L = dbInput.get();
+    LabelsH.at({N, L}) = 1;
+    // Even ResNet50 model got trained in NCHW format
+    for (unsigned C = 0; C < IMAGE_COLORS; ++C) {
+      auto BRGC = IMAGE_COLORS - 1 - C;
+      for (unsigned H = 0; H < 32; ++H) {
+        for (unsigned W = 0; W < 32; ++W) {
+          // ResNet BGR color space vs CIFAR RGB.
+          auto Val = static_cast<float>(static_cast<uint8_t>(dbInput.get()));
+          // Normalize and scale image.
+          Val = (Val - imagenetNormMean[C]) * Scale / imagenetNormStd[C] + Bias;
+          for (unsigned SH = 0; SH < SCALE; ++SH) {
+            for (unsigned SW = 0; SW < SCALE; ++SW) {
+              ImagesH.at({N, BRGC, SH + H * SCALE, SW + W * SCALE}) = Val;
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // From http://www.cs.toronto.edu/~kriz/cifar.html.
+  const char *TextualLabels[LABELS_SIZE] = {
+      "airplane", "automobile", "bird",  "cat",  "deer",
+      "dog",      "frog",       "horse", "ship", "truck"};
+
+  size_t SampleCounter = 0;
+  int ReportRate = 16;
+
+  llvm::Timer timer("Training", "Training");
+
+  llvm::outs() << "Training.\n";
+  for (size_t e = 0; e < numEpochs; ++e) {
+    llvm::outs() << "Training - epoch #" << e << " from total " << numEpochs
+                 << "\n";
+    // Train network on all examples.
+    timer.startTimer();
+    runBatch(EE, Bindings, ReportRate, SampleCounter, {A, E},
+             {&Images, &Labels});
+    timer.stopTimer();
+
+    llvm::outs() << "Evaluating.\n";
+    unsigned score = 0;
+    unsigned total = 0;
+
+    for (unsigned int i = 0; i < CIFAR_NUM_IMAGES / miniBatchSize; ++i) {
+      Tensor Sample(ElemKind::FloatTy, BatchImagesDims);
+      Sample.copyConsecutiveSlices(&Images, miniBatchSize * i);
+      updateInputPlaceholders(Bindings, {A}, {&Sample});
+      EE.run(Bindings);
+
+      for (unsigned int Batch = 0; Batch < miniBatchSize; ++Batch) {
```

Comment:
I think `Batch` is a poor name. Consider `Ex`, `Example` or `MiniBatchPos` or something similar.